Update how we Save and Propagate Privacy Preferences [#3013]

pattisdr opened the pull request:
❗ Contains migration; check downrev before merge.
Closes #3013 

### Code Changes

* [ ] Add two new models: `CurrentPrivacyPreference` (which is similar to the `Consent` table to which we will soon stop writing) and `PrivacyPreferenceHistory`.  
    * [ ] Every time a user saves preferences, we will create a `PrivacyPreferenceHistory` record for each preference which has a FK to a `PrivacyNoticeHistory` record.  We link to the historical notice as notices are subject to change over time.  
    * [ ] We then upsert `CurrentPrivacyPreferences` for the given user in a separate table for the given `PrivacyNotice`
    * [ ] A `PrivacyRequest` is then created, and each `PrivacyPreferenceHistory` record that was just created is given a FK to that`PrivacyRequest`. Request propagation will retrieve these PrivacyPreferenceHistory records, again linking to the historical record here, as the `CurrentPrivacyPreference` record is subject to change.
    * [ ] `PrivacyPreferenceHistory` is largely useful for propagation as well as for consent reporting. `CurrentPrivacyPreference` contains the user's latest saved preference for a given notice. `CurrentPrivacyPreference is separated for ease of querying and display of current preferences.
 * [ ] Largely leave the old workflow untouched for now (where we save consent preferences with respect to data uses defined in the privacy center).  However, start queueing a privacy request each time preferences are saved even if no preferences are executable.
 * [ ] Add new endpoint `consent_request_verify_for_privacy_preferences`  which allows you to verify your identity to retrieve your current current preferences (returns records from the `CurrentPrivacyPreference` table). This is similar to the existing endpoint that instead returns records from the `Consent` table.
 
**POST `{{host}}/consent-request/{{consent_request_id}}/verify-for-privacy-preferences`**
 ```json
{
  "code": "{{verification_code}}"
}
```
 * [ ] Add new endpoint `save_privacy_preferences `  which allows you to save privacy preferences with respect to privacy notice history records.  This creates `PrivacyPreferenceHistory` records, upserts `CurrentPrivacyPreferences` records, creates a PrivacyRequest to propagate, then returns the `CurrentPrivacyPreferences` that were in the request body.
 
**PATCH `{{host}}/consent-request/{{consent_request_id}}/privacy-preferences`**
```json
{
   "browser_identity": {
       "ga_client_id": "UA-XXXXXXXXX",
       "ljt_readerID": "test_sovrn_id"
   },
   "code": "{{verification_code}}",
   "preferences": [{
       "privacy_notice_history_id": "{{privacy_notice_history_id}}",
       "preference": "opt_out"
   }],
   "request_origin": "privacy_center",
   "url_recorded": "example.com/privacy_center",
   "user_agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_14_2)...",
   "user_geography": "us_ca"
}
```
* [ ] Return the privacy notice history id alongside the endpoints that get and create privacy notices so this is available when setting privacy preferences respect to those historical notices.
* [ ] Update the consent email template to handle both old (consent) and new style  (privacy preferences) being fed into it.
* [ ] Start adding skipped execution logs in both the consent and erasure email connectors when we opt not to send an email (for many reasons) for better record keeping
* [ ] Add shared logic to both consent email send and saas connector run_consent_request to be more discerning about whether the current preferences are applicable to the given system
* [ ] Adjust so saas connectors are only firing at most one series of requests: right now, either the entire opt in or the entire opt out series.  This was achieved previously by restricting so only one executable preference would be on the connector.

### Steps to Confirm

* See [Testing Notes Here](https://ethyca.atlassian.net/wiki/spaces/EN/pages/2657550453/Testing+Saving+and+Propagating+Privacy+Preferences+with+Privacy+Notices+Backend). 
* Big picture, the old workflow for saving consent should still work largely the same.  The new workflow should work too although it is partially API-only driven at the moment.

### Pre-Merge Checklist

* [ ] All CI Pipelines Succeeded
* Documentation:
  * [ ] documentation complete, [PR opened in fidesdocs](https://github.com/ethyca/fidesdocs/pulls)
  * [ ] documentation [issue created in fidesdocs](https://github.com/ethyca/fidesdocs/issues/new/choose)
* [ ] Issue Requirements are Met
* [ ] Relevant Follow-Up Issues Created
* [ ] Update `CHANGELOG.md`
* [x] For API changes, the [Postman collection](https://github.com/ethyca/fides/blob/main/docs/fides/docs/development/postman/Fides.postman_collection.json) has been updated

### Description Of Changes

Primary changes here:

- Start saving privacy preferences with respect to the "privacy notices"
    - Start storing a snapshot of the privacy preferences that were changed every single time they are saved for reporting purposes
    - Separately, have a new table that stores the most recently saved privacy preference for a given notice for ease of displaying a user's current preferences
- Expose endpoints to retrieve and save the new privacy preferences after verifying your identity (similar to the consent counterparts that are soon to be obsolete)
- For both the old workflow and the new workflow introduced here, start creating privacy requests every single time preferences are saved.
- Start adding restrictions as to whether we fire requests for a particular consent saas connector or send consent emails: in particular, start comparing privacy notice history data uses and system data uses. the privacy notice history must be enforceable system wide.
- Start adding skipped execution logs for a connector when we don't fire a request (instead of completed logs) and start adding skipped execution logs instead of no execution logs when we opt to not send an email for a connector for clearer record-keeping
- Note that the old workflow where we save consent with respect to data uses defined in the privacy center config.json is still preserved here for the time being.

pattisdr commented on src/fides/api/ops/models/privacy_preference.py:
```diff
@@ -0,0 +1,206 @@
+    privacy_modal = "privacy_modal"
+
+
+class PrivacyPreferenceHistory(Base):
+    """The DB ORM model for storing PrivacyPreferenceHistory, used for saving
+    every time consent preferences are saved for reporting purposes.
+
+    add fides_user_provided_identity nullable FK
+    add fides_user_device_id encrypted field
+    add ip address
+    add consent preference method?
+    """
```
This new `PrivacyPreferenceHistory` record should be created every single time a preference is saved for consent reporting.  It also is the primary record used in consent propagation.    The main goal here is to be storing as much relevant information as possible here to reduce the number of joins for reporting.

There are other fields that I need to add to this record for reporting, but as I need to get clarity on some of them, I will further expand this table as part of that work.

pattisdr commented on src/fides/api/ops/models/privacy_preference.py:
```diff
@@ -0,0 +1,206 @@
+        return created_privacy_preference_history
+
+
+class CurrentPrivacyPreference(Base):
+    """Stores only the user's most recently saved preference for a given privacy notice
+
+    The specific privacy notice history and privacy preference history record are linked as well.
+    """
+
+    preference = Column(EnumColumn(UserConsentPreference), nullable=False, index=True)
+    provided_identity_id = Column(
+        String,
+        ForeignKey(ProvidedIdentity.id),
+        nullable=True,
+    )
+    privacy_notice_id = Column(
+        String, ForeignKey(PrivacyNotice.id), nullable=False, index=True
+    )
+    privacy_notice_history_id = Column(
+        String, ForeignKey(PrivacyNoticeHistory.id), nullable=False, index=True
+    )
+    privacy_preference_history_id = Column(
+        String, ForeignKey(PrivacyPreferenceHistory.id), nullable=False, index=True
+    )
+
+    privacy_notice_history = relationship(PrivacyNoticeHistory)
+    privacy_preference_history = relationship(
+        PrivacyPreferenceHistory, cascade="delete, delete-orphan", single_parent=True
+    )
+
+    UniqueConstraint(
+        provided_identity_id, privacy_notice_id, name="identity_privacy_notice"
+    )
```
The main purpose of this table is to store the user's Current Privacy Preferences - it stores the latest preference saved for a unique privacy notice, not privacy notice history, although we store the history here too so it's clear which one they consented to. This table is analogous to the `Consent` table and was created to avoid messy queries on the bulky `PrivacyPreferenceHistory` table just to get current preferences.

pattisdr commented on src/fides/api/ops/service/connectors/consent_email_connector.py:
```diff
@@ -84,6 +94,34 @@ def test_connection(self) -> Optional[ConnectionTestStatus]:
                             Consent(data_use="advertising", opt_in=False),
                             Consent(data_use="improve", opt_in=True),
                         ],
+                        privacy_preferences=[
+                            MinimalPrivacyPreferenceHistorySchema(
+                                preference=UserConsentPreference.opt_in,
+                                privacy_notice_history=PrivacyNoticeHistorySchema(
+                                    name="Targeted Advertising",
+                                    regions=["us_ca"],
+                                    id="test_1",
+                                    privacy_notice_id="12345",
+                                    consent_mechanism=ConsentMechanism.opt_in,
+                                    data_uses=["advertising.first_party.personalized"],
+                                    enforcement_level=EnforcementLevel.system_wide,
+                                    version=1.0,
+                                ),
+                            ),
+                            MinimalPrivacyPreferenceHistorySchema(
+                                preference=UserConsentPreference.opt_out,
+                                privacy_notice_history=PrivacyNoticeHistorySchema(
+                                    name="Analytics",
+                                    regions=["us_ca"],
+                                    id="test_2",
+                                    privacy_notice_id="67890",
+                                    consent_mechanism=ConsentMechanism.opt_out,
+                                    data_uses=["improve.system"],
+                                    enforcement_level=EnforcementLevel.system_wide,
+                                    version=1.0,
+                                ),
+                            ),
+                        ],
```
This is the code that sends a test email for a saas email connector.  I'm putting in new-style privacy preferences into the email as well as leaving old style consent preferences.

pattisdr replied:
good idea!  I don't think I need to mark all of them, but this is one that could be easily forgotten I think.

pattisdr commented on src/fides/api/ops/service/connectors/consent_email_connector.py:
```diff
@@ -97,20 +135,48 @@ def test_connection(self) -> Optional[ConnectionTestStatus]:
     def needs_email(
```
`Saasconnector.needs_email` fires on the privacy request runner when we check if a privacy request needs to exit in a state of `awaiting_email_send` because we batch these emails and only send them once a week. I've modified whether we need to send a consent email to add in all the new checks, see: `filter_privacy_preferences_for_propagation`.  Note that the old workflow is still maintained here.

pattisdr commented on src/fides/api/ops/service/connectors/consent_email_connector.py:
```diff
@@ -97,20 +135,48 @@ def test_connection(self) -> Optional[ConnectionTestStatus]:
+            },
         )
 
     def batch_email_send(self, privacy_requests: Query) -> None:
```
batch_email_send is run once weekly and takes all consent privacy requests in a state of `awaiting_email_send` and batches relevant preferences together from all the privacy requests and sends in one email.

This also filters relevant preferences for this connector with `filter_privacy_preferences_for_propagation`.  Unlike the saas connector, if there are conflicting preferences, I send them both here.

pattisdr commented on src/fides/api/ops/service/connectors/consent_email_connector.py:
```diff
@@ -119,15 +185,49 @@ def batch_email_send(self, privacy_requests: Query) -> None:
                 )
             else:
                 skipped_privacy_requests.append(privacy_request.id)
+                ExecutionLog.create(
+                    db=db,
+                    data={
+                        "connection_key": self.configuration.key,
+                        "dataset_name": self.configuration.name,
+                        "collection_name": self.configuration.name,
+                        "privacy_request_id": privacy_request.id,
+                        "action_type": ActionType.consent,
+                        "status": ExecutionLogStatus.skipped,
+                        "message": f"Consent email skipped for '{self.configuration.name}'",
+                    },
+                )
```
For better record-keeping, we were just not adding any logs for email connectors if the request wasn't fired. Now adding a skipped log instead.  Actually, I forgot I made this a method above, so I'll call `self.add_skipped_log(db, privacy_request)` here

pattisdr commented on src/fides/api/ops/service/connectors/erasure_email_connector.py:
```diff
@@ -104,6 +105,7 @@ def batch_email_send(self, privacy_requests: Query) -> None:
                 batched_identities.extend(filtered_user_identities.values())
             else:
                 skipped_privacy_requests.append(privacy_request.id)
+                self.add_skipped_log(db, privacy_request)
```
Making the corresponding skipped log change for the erasure connector @galvana

pattisdr commented on src/fides/api/ops/service/connectors/saas_connector.py:
```diff
@@ -427,56 +432,76 @@ def run_consent_request(
         )
         self.set_privacy_request_state(privacy_request, node)
         query_config = self.query_config(node)
+        should_opt_in: Optional[bool] = should_opt_in_to_service(
+            self.configuration.system, privacy_request
+        )
```
I'm having the saas connector collapse all the privacy preferences and/or consent preferences (old style) into a single optional boolean.  if True, we "opt in". If False, we "opt out". If "None" we do nothing.  see `should_opt_in_to_service` util which uses the same `filter_privacy_preferences_for_propagation` that the consent email connector uses in determining whether a preference is relevant for the current connector.

pattisdr commented on src/fides/api/ops/service/connectors/saas_connector.py:
```diff
@@ -427,56 +432,76 @@ def run_consent_request(
         policy: Policy,
         privacy_request: PrivacyRequest,
         identity_data: Dict[str, Any],
-        consent_preferences: List[Consent],
     ) -> bool:
         """Execute a consent request. Return whether the consent request to the third party succeeded.
-        Return True if 200 OK
+        Should only propagate either the entire set of opt in or opt out requests.
+        Return True if 200 OK. Raises a SkippingConsentPropagation exception if no action is taken
+        against the service.
```
The bulk of the code in this SaaSConnector.run_consent_request is largely unchanged - there's one degree less of indentation because I don't loop through all the preferences, they are collapsed into a single optional boolean ahead of time as to whether or not we will fire the request and which set of opt-in/opt-out preferences we will propagate.

pattisdr commented on src/fides/api/ops/service/connectors/saas_connector.py:
```diff
@@ -427,56 +432,76 @@ def run_consent_request(
+                node.address.value,
+                query_config.action,
+            )
+            raise SkippingConsentPropagation(
+                f"Skipping consent propagation for node {node.address.value} -  No '{query_config.action}' requests defined."
+            )
```
Raising SkippingConsentPropagation in multiple places on this connector, which actually just creates a "skipped" log here instead of a "completed" log for this node - handled in GraphTask.

pattisdr commented on src/fides/api/ops/service/privacy_request/request_runner_service.py:
```diff
@@ -684,15 +684,55 @@ def needs_batch_email_send(
+        else:
+            can_skip_consent_email.append(connection_config)
+
+    if not needs_email_send:
+        _create_execution_logs_for_skipped_email_send(
+            db, privacy_request, can_skip_erasure_email, can_skip_consent_email
+        )
```
In the request runner, if we determine there are no emails to send at all, but there are consent or erasure email connectors configured, add skipped logs for every email connector here.   Otherwise, we'll wait until the weekly email send to add the logs.

pattisdr commented on src/fides/api/ops/task/graph_task.py:
```diff
@@ -105,7 +102,11 @@ def result(*args: Any, **kwargs: Any) -> Any:
+                except (
+                    CollectionDisabled,
+                    NotSupportedForCollection,
+                    SkippingConsentPropagation,
+                ) as exc:
```
This is in the wrapper that wraps nodes in the graph that also creates the execution logs - here is where raising a SkippingConsentPropagation in execution will create a skipped log for that node.

pattisdr commented on src/fides/api/ops/task/graph_task.py:
```diff
@@ -587,16 +588,11 @@ def consent_request(self, identity: Dict[str, Any]) -> bool:
             )
             return False
 
-        consent_preferences: List[Consent] = [
-            Consent(**pref) for pref in self.resources.request.consent_preferences or []
-        ]
-
         output: bool = self.connector.run_consent_request(
             self.traversal_node,
             self.resources.policy,
             self.resources.request,
             identity,
-            consent_preferences,
```
These are the old style consent preferences. I'm just pulling them off of the privacy request inside saasconnector.run_consent_request instead of passing them in here as a first step to getting rid of this old workflow in the future.

pattisdr commented on src/fides/api/ops/util/consent_util.py:
```diff
@@ -0,0 +1,80 @@
+    return _filter_matching_system_data_uses(system, propagatable_preferences)
+
+
+def should_opt_in_to_service(
+    system: Optional[System], privacy_request: PrivacyRequest
+) -> Optional[bool]:
+    """
+    Determine if we should opt in (return True), opt out (return False), or do nothing (return None) for the given System.
+
+    - If using the old workflow (privacyrequest.consent_preferences), return True if all attached consent preferences
+    are opt in, otherwise False.  System check is ignored.
+
+    - If using the new workflow (privacyrequest.privacy_preferences), there is more filtering here.  Privacy Preferences
+    must have an enforcement level of system-wide and a data use must match a system data use.  If the connector is
+    orphaned (no system), skip the data use check. If conflicts, prefer the opt-out preference.
```
This file has the main logic for whether preferences should be propagated for both saas and consent email connectors.

pattisdr commented:
@adamsachs here is the rewrite of https://github.com/ethyca/fides/pull/2991 based on [discussions Friday](https://ethyca.atlassian.net/wiki/spaces/EN/pages/2661777424/2023-04-07+Proposed+Changes+to+how+we+Save+Consent+Preferences).  Would you have time to review? This PR looks bigger than it is, but a lot of the changes are new tests.

@seanpreston I'm tagging you here because this new direction we're going with consent reporting (new tables created/new endpoints) likely won't solve your issue of surfacing the old-style consent records.

adamsachs reviewed (commented):
i'm not through this fully yet, but wanted to post an incremental review before i end my day here, so that - if you want - you can start to address any comments (none of them major so far!) in the meantime. i haven't done any testing yet, or looked over the automated test updates, so that's one big area i need to get to, and some additional code comments may come from that, as i get a bit more hands-on. i've also not dug into the email connector changes - just because that felt like a somewhat independent chunk that i could wall off for the time being - but i will look to review those ASAP tomorrow.

what i have looked over - which i think gets to the core of the changes in the PR - generally looks really good to me! i don't see any major issues so far. 

there are two categories of optimizations that i wanted to entertain: one questioning whether we should be indexing some more db columns, and another about whether it'd be possible to DRY up some of the matching we do between privacy notices and systems. the former is obviously something we should look into now (though i very well may be misguided in my points), while the latter doesn't need to be addressed in this PR by any means. there are a few other minor points throughout, but they should be easily resolved or dismissed. 

i'll keep on looking at this as fast as i can, but it's looking very promising!

pattisdr commented on src/fides/api/ctl/sql_models.py:
```diff
@@ -337,6 +337,18 @@ def get_system_data_uses(
                         data_uses.add(data_use)
         return data_uses
 
+    def get_data_uses(self, include_parents: bool = True) -> Set[str]:
```
For sure, thanks for highlighting this.

pattisdr commented on src/fides/api/ops/models/privacy_preference.py:
```diff
@@ -0,0 +1,206 @@
+        ForeignKey(ProvidedIdentity.id),
+        nullable=True,
+    )
+    relevant_systems = Column(
+        MutableList.as_mutable(ARRAY(String))
+    )  # Systems whose data use match.  This doesn't necessarily mean we propagate. Some may be intentionally skipped later.
+    request_origin = Column(EnumColumn(RequestOrigin))
```
Just missing some cleanup, thanks for this.

`nullable=True` is the default so if I'm missing it, I probably knew it wasn't required for the onset. Some where I've added nullable=True explicitly, it's possible I changed them from False->True, as I originally had marked too many fields as non-nullable, or I copied a field off another model.

I think it's easier to parse if only the fields that are required have nullable=False, otherwise that argument is left off altogether.

pattisdr commented on src/fides/api/ops/models/privacy_preference.py:
```diff
@@ -0,0 +1,206 @@
+    affected_system_status = Column(
+        MutableDict.as_mutable(JSONB), server_default="{}", default=dict
+    )
+    email = Column(
```
You're right I hadn't added an index because it was encrypted, but I see on the ProvidedIdentity table we add an index to the hashed value instead so I can probably add two new columns for that here.

pattisdr commented on src/fides/api/ops/api/v1/endpoints/privacy_preference_endpoints.py:
```diff
@@ -0,0 +1,217 @@
+
+    query = (
+        db.query(CurrentPrivacyPreference)
+        .join(
```
Good catch, this wasn't an intentional join, I assume it's just the remnants of my original nightmare of a query here that I built on the `PrivacyPreferenceHistory` table that led me to create `CurrentPrivacyPreference` in the first place.

It should be faster to filter on provided_identity_id like you suggest and I'll add an index to provided_identity_id while I'm here, that looks like an oversight too.

pattisdr commented on src/fides/api/ops/api/v1/endpoints/privacy_preference_endpoints.py:
```diff
@@ -0,0 +1,217 @@
+    db: Session, data: PrivacyPreferencesCreateWithCode
+) -> None:
+    """
+    Runs a check that makes sure all privacy notice histories exist and also makes sure we don't have preferences specified
+    for the same privacy notice when saving privacy preferences.
+
+    For example, we want to avoid having two preferences saved for the same version of a *historical privacy notice*,
+    or two preferences saved for different versions of the same *privacy notice*.
```
Thanks for the suggestion!

pattisdr commented on src/fides/api/ops/models/privacy_preference.py:
```diff
@@ -0,0 +1,206 @@
+        existing_current_preference = (
+            db.query(CurrentPrivacyPreference)
+            .filter(
+                CurrentPrivacyPreference.provided_identity_id
```
Yes thank you! I do need an index on that column.  I don't think postgres automatically creates an index for a FK column.

pattisdr commented on src/fides/api/ops/models/privacy_preference.py:
```diff
@@ -0,0 +1,206 @@
+    )
+
+    @classmethod
+    def create(
```
What do you think of the decision to collapse CurrentPrivacyPreference records based on notice instead of notice history id? While we're saving privacy preferences generally against the historical record, in my opinion it made sense to make the notice unique here.  

Trying to think of what the UI would need to surface "current preferences".   Say the most current preferences we're displaying in the UI are an Advertising notice v6 but we only have preferences saved against v5.  I'd think the v5 preference is still more relevant than whatever the default preference is in that case.

pattisdr commented on src/fides/api/ops/util/consent_util.py:
```diff
@@ -0,0 +1,80 @@
+        and pref.preference != UserConsentPreference.acknowledge
+    ]
+
+    if not system:
```
I did run this by product, but if it stood out to you, let's make sure this is the right call.

If the connector is orphaned (the ConnectionConfig is not linked to a System) Michael said to preserve our current default behavior of firing all opt in or opt out behavior as we do now.  But then once we get rid of orphaned connectors, this issue will go away. 

While I agree with you that normally if the connector isn't associated with a system, we wouldn't want to fire, there's this other detail that ConnectionConfigs and Systems were traditionally separate concepts in separate codebases for a long time.  To me, our UI still promotes creating them as separate things.

adamsachs replied:
OK - works for me!

pattisdr commented on src/fides/api/ops/util/consent_util.py:
```diff
@@ -0,0 +1,80 @@
+    for pref in preferences:
+        privacy_notice_data_uses: List[str] = pref.privacy_notice_history.data_uses or []  # type: ignore[union-attr]
+        for privacy_notice_data_use in privacy_notice_data_uses:
+            for system_data_use in system.get_data_uses(include_parents=True):
```
I was thrilled all your work got merged last week.  Very handy to be able to use this on the rewrite.

pattisdr commented on src/fides/api/ops/util/consent_util.py:
```diff
@@ -0,0 +1,80 @@
+        for privacy_notice_data_use in privacy_notice_data_uses:
+            for system_data_use in system.get_data_uses(include_parents=True):
+                if system_data_use == privacy_notice_data_use:
+                    filtered_preferences.append(pref)
```
I've consolidated two of the locations here! Thanks for the suggestion.

pattisdr commented on src/fides/api/ops/models/privacy_notice.py:
```diff
@@ -239,4 +258,19 @@ class PrivacyNoticeHistory(PrivacyNoticeBase, Base):
+                        if system_data_use == privacy_notice_data_use:
+                            relevant_systems.append(system.fides_key)
+                            continue
+        return relevant_systems
```
What I've done so far is try to add this new `PrivacyNotice.applies_to_system(system)` that I am using where I go to cache relevant systems on the PrivacyPreferenceHistory record, and where I go to later filter privacy preferences for propagation. 

 I could also use this in your privacy notice endpoints, but you've done other things to make this more efficient over there - with your notices by data use lookup for example.

adamsachs reviewed (commented):
ok, went through the latest changes (look great generally!) and wanted to publish another round of comments just so you can get that feedback ASAP.

getting to testing now...

adamsachs commented on docs/fides/docs/development/postman/Fides.postman_collection.json:
```diff
@@ -1,8 +1,8 @@
+		"_postman_id": "5d1a5ee9-f432-4da2-a9c9-d68c17e776ac",
 		"name": "Fides",
-		"schema": "https://schema.getpostman.com/json/collection/v2.1.0/collection.json"
+		"schema": "https://schema.getpostman.com/json/collection/v2.1.0/collection.json",
```
looks like theres an additional `,` here that's causing parsing errors when i try to import into postman. just removing the trailing `,` made things work.

adamsachs commented on tests/ops/models/test_privacy_preference.py:
```diff
@@ -0,0 +1,172 @@
+                check_name=False,
+            )
+
+    def test_create_privacy_notice(self, db, privacy_notice, system, privacy_request):
```
think the method name just needs to be updated here :)

adamsachs commented on tests/ops/util/test_consent_util.py:
really nice tests here, super easy to follow!

pattisdr commented on tests/ops/service/connectors/test_saas_connector.py:
```diff
@@ -423,3 +426,192 @@ def test_get_consent_requests_by_preference(
+        privacy_preference_history,
+        db,
+    ):
+        """User is expressing an opt in preference here but GA only has an opt out preference defined"""
```
sure thing! this behavior isn't new by the way and is end-to-end tested in the google analytics connector but I thought it would be good to add some more intermediate tests to this method while I'm here (also codecov likely flagged this if I recall)

adamsachs replied:
oh yeah, i just meant all the docstrings in your test methods generally :)

pattisdr replied:
oh good!

pattisdr commented on src/fides/api/ops/api/v1/endpoints/consent_request_endpoints.py:
```diff
@@ -260,20 +265,14 @@ def queue_privacy_request_to_propagate_consent(
         if pref.data_use in executable_data_uses
     ]
 
-    if not executable_consent_preferences:
-        logger.info(
-            "Skipping propagating consent preferences to third-party services as "
-            "specified consent preferences: {} are not executable.",
-            [pref.data_use for pref in consent_preferences.consent or []],
-        )
-        return None
-
```
Reticketed here @seanpreston

pattisdr commented:
OK I think I've addressed your comments @adamsachs and have completed another round of manual testing.  A few new fields were added or adjusted based on changing requirements.

adamsachs reviewed (approved):
@pattisdr i think this is ready to go from my POV!

thanks for the diligence in addressing all the comments. and 👏 for turning around a huge enhancement in such quick time, in a clean and easy-to-follow manner!

pattisdr commented:
Likewise @adamsachs, appreciate the detailed, prompt review! ⭐
